### PR TITLE
DrivAerML: WD+gc compound at champion LR — gentle regularization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The champion DM config has NO gc and NO WD. #3011 showed wd=5e-3 is too far from wd=1e-2, and #3010 showed gc=0.5/2.0 don't help alone. But what about GENTLE compounds? Test wd=1e-3 (1/10 of what AF uses) + gc=1.0 as a mild regularization overlay on the champion config. The idea: the DM NaN cliff at ~ep178 might be avoidable with very light regularization that doesn't disturb the optimizer trajectory.

## Current Baselines

| Dataset | Metric | Value |
|---------|--------|-------|
| TF | val_primary/surface_pressure_mae | 22.537 |
| TFP | val_primary/field_mse | 0.00434 |
| AF | val_primary/surface_mse | 0.000482 |
| DM | val_primary/surface_rel_l2_pct | 3.997% |
| DM external target | val_primary/surface_rel_l2_pct | 3.71% |

## Runs

Base command (all runs):
```bash
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group dm_gentle_reg
```

**Run 1** — very light WD only (wd=1e-3, no gc):
```bash
<base> --weight-decay 1e-3
```

**Run 2** — light compound (wd=1e-3 + gc=1.0):
```bash
<base> --weight-decay 1e-3 --grad-clip 1.0
```

**Run 3** — pure clip test at champion config (gc=1.0, no WD):
```bash
<base> --grad-clip 1.0
```

**Run 4** — champion control with full eval (no --max-eval-batches):
```bash
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --wandb_group dm_gentle_reg
```

## What to Report

For each run: best `val_primary/surface_rel_l2_pct`, epoch achieved, and W&B run ID. Note if any NaN divergence is observed and at which epoch.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT sukuna:

## Results

### Summary

| Run | Config | Best Val `surface_rel_l2_pct` | Best Test `surface_rel_l2_pct` | Best Epoch | Total Epochs | W&B ID | Stability |
|-----|--------|------|------|------------|-------------|--------|-----------|
| 1 | wd=1e-3, no gc | 22.07% | 23.40% | 31 | 449 | `9hflnw2d` | **Diverged ~ep100** (grad_norm → millions) |
| 2 | wd=1e-3 + gc=1.0 | **4.44%** | **5.30%** | 421 | 448 | `zromiqrc` | **Stable throughout** |
| 3 | gc=1.0 only, no wd | 5.20% | 6.27% | 229 | 443 | `imyzfshf` | **Diverged ~ep250** (grad_norm → inf) |
| 4 | champion (full eval) | 17.67% | 17.05% | 31 | 44 | `igbb8kb2` | Too few epochs (full eval ~10x slower) |

**Baseline:** val_primary/surface_rel_l2_pct = 3.997% | External target = 3.71%

### Key Findings

1. **WD alone (Run 1) diverges completely.** Gradient norms explode after ~epoch 100 (reaching millions by epoch 400). Best metric achieved before divergence was only 22.07% — far from competitive. WD=1e-3 alone is insufficient to stabilize the champion DM config.

2. **The compound WD+GC (Run 2) is the clear winner.** It was the only configuration that remained stable for the full training duration (448 epochs). Best val of 4.44% is the closest to baseline (3.997%), with test at 5.30%. Grad norms stayed well-controlled (~0.13 at late training).

3. **GC alone (Run 3) is initially strong but eventually diverges.** Reached 5.20% val (best at epoch 229) but then grad norms spiked and the run diverged around epoch 250. This shows gc=1.0 alone delays but doesn't prevent the instability — the weight decay provides a complementary regularization effect.

4. **Champion full eval (Run 4) is inconclusive** — only completed 44 epochs in the 6-hour timeout because evaluating the full eval set (no --max-eval-batches cap) makes each epoch ~10x slower. Its best of 17.67% at epoch 31 is not a meaningful comparison.

### Analysis

The hypothesis is partially confirmed: gentle regularization **does** stabilize the DM training trajectory. The compound (wd=1e-3 + gc=1.0) prevented the NaN cliff entirely, while neither component alone was sufficient for full-duration stability.

However, the compound at 4.44% val didn't beat the champion baseline of 3.997% — it's ~11% worse. This suggests the regularization, while stabilizing, slightly constrains the model's capacity to fit. The champion config (no WD, no gc) reaches a better optimum when it doesn't diverge, but at the cost of instability.

The stability ordering is clear: **WD+GC > GC only > WD only > neither** (in terms of how long before divergence).

### Commands Used

```bash
# Run 1: wd=1e-3, no gc
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group dm_gentle_reg --weight-decay 1e-3 --agent sukuna --wandb-name "sukuna/dm-wd1e3-no-gc"

# Run 2: wd=1e-3 + gc=1.0 (compound)
CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group dm_gentle_reg --weight-decay 1e-3 --grad-clip 1.0 --agent sukuna --wandb-name "sukuna/dm-wd1e3-gc1.0"

# Run 3: gc=1.0 only
CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group dm_gentle_reg --grad-clip 1.0 --agent sukuna --wandb-name "sukuna/dm-gc1.0-no-wd"

# Run 4: champion control (full eval, no --max-eval-batches)
CUDA_VISIBLE_DEVICES=3 SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --wandb-group dm_gentle_reg --agent sukuna --wandb-name "sukuna/dm-champion-full-eval"
```

### Peak Memory Usage
~6 GB per GPU (observed via nvidia-smi during training)

### Suggested Follow-ups

1. **Try the compound at the champion's full epoch budget** — Run 2 was still improving at epoch 421. With more training time, it might close the gap to 3.997%.
2. **Tune the gc threshold** — gc=1.0 is a reasonable first guess, but gc=0.5 or gc=2.0 with wd=1e-3 might find a better stability/performance tradeoff.
3. **Try wd=5e-4 + gc=1.0** — even lighter WD might retain more model capacity while still providing the synergistic stabilization observed here.
4. **Re-run Run 4 with --max-eval-batches 200** — the full eval control was too slow to be useful. Running the champion with the same eval budget would give a proper apples-to-apples baseline.

**Comment by tcapelle** [comment]
ADVISOR: WD=1e-3 + gc=1.0 at 4.44% is a promising stable result — the compound works and the gap to baseline is modest. Please follow your instinct and retry at **WD=5e-4 + gc=1.0** (half the regularization). Also run **WD=1e-4 + gc=1.0** as a second variant in the same PR if you have capacity. Goal: find the WD level where the stability from gc is preserved but the model has enough capacity to beat 3.997%.

**Comment by tcapelle** [comment]
STUDENT sukuna:

## Results — Follow-up: lighter WD compounds

Per advisor request, tested two lighter WD levels with gc=1.0 to find the sweet spot where stability is preserved but model capacity is maximized.

### Summary

| Run | Config | Best Val `surface_rel_l2_pct` | Best Test `surface_rel_l2_pct` | Best Epoch | Total Epochs | W&B ID | Stability |
|-----|--------|------|------|------------|-------------|--------|-----------|
| 2 (prev) | wd=1e-3 + gc=1.0 | **4.44%** | **5.30%** | 421 | 448 | `zromiqrc` | Stable throughout |
| 5 | wd=5e-4 + gc=1.0 | 12.02% | 12.71% | 47 | 500 | `kepvb37k` | **Diverged ~ep71** (grad_norm → inf) |
| 6 | wd=1e-4 + gc=1.0 | 4.63% | 5.42% | 370 | 501 | `owholcg1` | **Diverged ~ep499** (grad_norm → inf) |

**Baseline:** val_primary/surface_rel_l2_pct = 3.997% | External target = 3.71%

### Key Findings

1. **WD=5e-4 + gc=1.0 (Run 5) is surprisingly the worst.** It diverged early at epoch 71, achieving only 12.02% val before grad norms exploded. This is counterintuitively worse than both stronger (1e-3) and weaker (1e-4) weight decay.

2. **WD=1e-4 + gc=1.0 (Run 6) nearly matches the best compound.** Best val of 4.63% at epoch 370, compared to 4.44% for WD=1e-3. It stayed stable for ~499 epochs before eventually diverging — nearly as stable as WD=1e-3 which never diverged in 448 epochs.

3. **WD=1e-3 + gc=1.0 remains the best compound.** The original compound from Run 2 still holds the best val (4.44%) and best test (5.30%), with the most robust stability (no divergence at all during training).

### Analysis

The non-monotonic relationship between WD strength and performance is interesting. WD=5e-4 sits in a problematic middle ground — enough regularization to constrain the optimizer but not enough to prevent the gradient instability that causes divergence. At WD=1e-4, the weight decay is so light it barely affects the optimizer dynamics, letting gc=1.0 do the heavy lifting for stability (similar to gc-only but with a tiny stabilizing effect). At WD=1e-3, the compound provides robust dual-axis regularization.

None of the compound variants beat the baseline of 3.997%. The gap narrows (4.44% → 3.997% is ~11%) but doesn't close. The gentle regularization successfully prevents the NaN cliff but comes at a slight performance cost.

### Commands Used

```bash
# Run 5: wd=5e-4 + gc=1.0
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group dm_gentle_reg --weight-decay 5e-4 --grad-clip 1.0 --agent sukuna --wandb-name "sukuna/dm-wd5e4-gc1.0"

# Run 6: wd=1e-4 + gc=1.0
CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group dm_gentle_reg --weight-decay 1e-4 --grad-clip 1.0 --agent sukuna --wandb-name "sukuna/dm-wd1e4-gc1.0"
```

### Peak Memory Usage
~6 GB per GPU

### Suggested Follow-ups

1. **WD=1e-3 + gc=1.0 is the optimal compound** — further WD reduction doesn't help. The next lever would be tuning gc (e.g. gc=0.5 or gc=2.0 with WD=1e-3) to see if clip strength matters more than WD level.
2. **Cosine schedule extension** — the WD=1e-3 + gc=1.0 run was still improving at epoch 421. Extending cosine-t-max beyond 30 could let the optimizer explore longer at higher LR.
3. **The 3.997% baseline may be stochastic** — the champion config diverges on some seeds. The compound at 4.44% might be more reliable across seeds, which has practical value even if slightly worse on this metric.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3046 because ADVISOR: Closing — WD+gc compound best val=4.44% doesn't beat baseline (now 3.700%). These runs also lacked EMA which is now standard. Sukuna reassigned..
